### PR TITLE
Cache created judgement in `HitObject`

### PR DESCRIPTION
**Pull request by EVAST9919**
These values have been taken during beatmap loading (going from song-select to player loader and waiting until player  screen push)
|master|pr|
|---|---|
|master-loading|pr-loading|

### CODE REVIEW

**Comment by EVAST9919** [comment]
See https://github.com/ppy/osu/issues/27104#issuecomment-1936997683

**Comment by EVAST9919** [comment]
Reopening since it improves allocations during beatmap loading (OP has been updated), and potentially has other benefits.

File: osu.Game/Rulesets/Objects/HitObject.cs [hunk 1]
```diff
@@ -162,11 +162,19 @@ protected virtual void CreateNestedHitObjects(CancellationToken cancellationToke
...
+        public Judgement Judgement => judgement ??= CreateJudgement();
+
+        private Judgement judgement;
+
         /// <summary>
         /// Creates the <see cref="Judgement"/> that represents the scoring information for this <see cref="HitObject"/>.
         /// </summary>
         [NotNull]
-        public virtual Judgement CreateJudgement() => new Judgement();
+        protected virtual Judgement CreateJudgement() => new Judgement();
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/HitObject.cs:177]: I'd prefer if this doesn't change for now, to prevent breaking changes.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
I think this is fine for as long as `Judgement` is only used to ferry fixed data. Looking through a few rulesets, it doesn't look like implementations do anything other than override `MaxResult`, which is good because that isn't really supported anyway.

We'll change this class [in the future](https://github.com/ppy/osu/pull/26563), at which time I'd reconsider the semantics around it - likely making it not a class and potentially moving it to the ctor.

Curious what @bdach thinks of this PR.

**Comment by bdach** [comment]
I was deferring review of this as it seemed relatively low priority. Without looking the diff over, mixed thoughts because I'm scared of this going wrong and storing data it shouldn't be storing, but also thinking that if we're ever to actually fix issues in editor stemming from lack of frame stability without _relying_ on frame stability it'd be by storing stuff more permanently to hitobjects rather than rely on DHOs for storing object state, which this helps in. https://github.com/ppy/osu/issues/11980 is also semi-relevant in thix context.

**Comment by smoogipoo** [comment]
One other thing to consider - and bear in mind I'm just listing my thoughts and doesn't necessarily affect this PR - is that at some point we'll probably want to pool `HitObject`s as they're the single highest source of GC pressure affecting things like diffcalc.
Anything stateful inside `HitObject` should probably consider that it may become transient at some point so that we don't inadvertently lock ourselves into more breaking changes further down the track

**Comment by peppy** [comment]
the judgement class really needs very strong xmldoc saying that it shouldn't be storing state. or it should be forcefully sealed from adding new fields if that's even a thing we can do.

**Comment by smoogipoo** [comment]
> or it should be forcefully sealed from adding new fields

Unfortunately not 😢 

But yeah I would see it going in a `struct` direction, it effectively sealing it. Something like:
```
ctor()
{
    Criteria = new JudgementCriteria
    {
        MaxResult = LargeTickHit,
        [MinResult = ...]
    }
}
```

**Comment by peppy** [comment]
Sounds good.

**Review by peppy — APPROVED** [review]

**Pull request merged**